### ISSUE TITLE
support context.Context arguments

### ISSUE THREAD

**Issue by srenatus** [keygen-sh/keygen-go#14]
Hey there.

The SDK does a lot of API interactions with the keygen service, and it would thus be nice if it was possible to abort them on timeouts -- via the standard context cancellation mechanism used in golang.

I have not found a way to do that with the current API, as none of its methods accept a context.Context parameter. Am I missing something?

Thanks! 😃

**Comment by ezekg** [comment]
What would the ideal API look like here?

**Comment by srenatus** [comment]
It might be v3 material, but I'd imagine that every function call that ends up doing an HTTP request would include a context.Context (first) argument. You wouldn't even need to do much more, the stdlib `http.Client` will deal with the rest.

Concretely,
```go
// Get is a convenience helper for performing GET requests.
func (c *Client) Get(path string, params interface{}, model interface{}) (*Response, error) {
	req, err := c.new(http.MethodGet, path, params)
	if err != nil {
		return nil, err
	}

	return c.send(req, model)
}
```
⏬ 
```go
// Get is a convenience helper for performing GET requests.
func (c *Client) Get(ctx context.Context, path string, params interface{}, model interface{}) (*Response, error) {
	req, err := c.new(ctx, http.MethodGet, path, params)
	if err != nil {
		return nil, err
	}

	return c.send(req, model)
}
```
and in `new(..)`, we'd use [`http.NewRequestWithContext`](https://pkg.go.dev/net/http#NewRequestWithContext). Furthermore, the calls to `client.Get()` (etc), like
```go
func (l *License) Machine(id string) (*Machine, error) {
```
would become
```go
func (l *License) Machine(ctx context.Context, id string) (*Machine, error) {
```

**Comment by ezekg** [comment]
That makes the API worse for people who don't need a context. Maybe we can add a new set of methods, e.g. `license.ActivateWithContext(ctx, ...)`. But then that makes the API surface area much larger, and also harder to maintain.

What's your use case for using context here? Timeout configuration can be handled using a custom HTTP client, so curious what else you're trying to do here.

**Comment by srenatus** [comment]
My use case is cancellation. For example, we've got a shutdown routine in our service, and there's a timeout involved: the time the application is granted for a clean shutdown. It's driven by a context with a timeout, that's passed along to all the subsystems that need to be shut down. Golang's context.Context is a good convention for that, in my opinion.

There are also cancellation scenarios where it's not driven by a timeout: say, there could be an HTTP handler that's calling one of keygen-go's methods. If the caller goes away, the request context is cancelled, and no needless work -- results that nobody is interested in anymore -- would have to be done, it would all be cancelled. Without the context argument, the call to keygen's APIs would need to finish regardless of the results being thrown away. This scenario isn't possible with a timeout, because you cannot know _when_ the caller goes away.

Now, it might be said that it's a contrived example; but the point I'm trying to make is that someone using Go often just expects the ctx argument. And having it wired up automatically allows all the use cases people *could have*.

> That makes the API worse for people who don't need a context. 

I might be wrong, but I think anyone using Golang for some time is rather expecting a context argument for a method that's crossing an network boundary than be annoyed that there's a context if they don't care about it 😅 

But it's fine, if you feel differently on this, people who need this can still create their own methods. Since we don't need all of the API surface, it's not much to replace.

**Comment by ezekg** [comment]
For backwards compatibility, I'd be open to doing something like this pseudo code:

```go
func (c *Client) GetWithContext(ctx context.Context, path string, params interface{}, model interface{}) (*Response, error) {
  req, err := c.new(ctx, http.MethodGet, path, params)
  if err != nil {
    return nil, err
  }

  return c.send(req, model)
}

func (c *Client) Get(path string, params interface{}, model interface{}) (*Response, error) {
  return GetWithContext(context.Background(), path, params, model)
}
```

Would something like that work?

**Comment by srenatus** [comment]
It would work, but we'd need to deal with all its callers, too... For example, https://github.com/keygen-sh/keygen-go/blob/fe7694901250ab90b1f37d35bd927b7e0f2305b4/license.go#L208-L219
would also need to become
```go
// Machines lists up to 100 entitlements for the license.
func (l *License) Entitlements() (Entitlements, error) {
	return l.EntitlementsWithContext(context.Background())
}

func (l *License) EntitlementsWithContext(ctx context.Context) (Entitlements, error) {
	client := NewClient()
	entitlements := Entitlements{}

	if _, err := client.GetWithContext(ctx, "licenses/"+l.ID+"/entitlements", querystring{Limit: 100}, &entitlements); err != nil {
		return nil, err
	}

	return entitlements, nil
}
```

So, it's less than ideal. But I also fail to come up with a better idea just now. 😅 

It would be possible to pass something like
```go
type contextThing struct {
  wrapped any
  ctx context.Context
}

func WithContext(ctx context.Context, x any) any {
  return contextThing{ctx: ctx, wrapped: x}
}
```
in `client.Get("...", WithContext(ctx, ...), ...)` to opportunistically context-ify something, but that would still need two methods for `(*License).Entitlements`. 🤔

**Comment by mathnogueira** [comment]
I prefer to move the API to a v3, but if that's out of the cogitation, I think duplicating the methods and providing a `DoSomethingWithContext` variant would be the best approach. 

But I agree this is a feature that is needed - at least for my team - to add tracing to the keygen client.

**Comment by ezekg** [comment]
@mathnogueira I'm fine with adding context to all methods for v3. Feel free to open a PR.

Useful discussion and implementation: https://github.com/google/go-github/issues/526

**Comment by ezekg** [comment]
Closed in #18. Now available in v3.0.0.
